spring-cloud-gcp-docs/src/main/asciidoc/core.adoc

```diff
@@ -135,6 +135,35 @@ and append the additional scopes you need to add.
 spring.cloud.gcp.credentials.scopes=DEFAULT_SCOPES,https://www.googleapis.com/auth/cloud-vision
 ----
 
-==== Spring Initializr
+=== Environment
+
+`GcpEnvironmentProvider` is a functional interface, auto-configured by the Spring Cloud GCP starter, that returns a `GcpEnvironment` enum.
+The provider can help determine programmatically in which GCP environment (App Engine Flexible, App Engine Standard, Kubernetes Engine or Compute Engine) the application is deployed.
+
+[source,java]
+----
+public interface GcpEnvironmentProvider {
+	GcpEnvironment getCurrentEnvironment();
+}
+----
+
+=== ResourceManager
+
+A default instance of Google Cloud `ResourceManager` is auto-configured by the Spring Cloud GCP starter.
+
+
+=== Environment
+
+`MetadataProvider` is another functional interface that provides access to Google Cloud Metadata Server attributes.
+The implementation auto-configured by the Spring Cloud GCP starter is a thin wrapper around static `MetadataConfig` utility, which is mainly helpful for testability.
```

Comment:
Why do we need to pollute production code with auto-configured beans for testing?

Can't we just live without it and really use only in tests if that?

Reply:
The client library utility for accessing metadata server is a static utility method on MetadataConfig class. I prefer to stay away from using static utilities, since they make the code using them inflexible and hard to test, so I made a wrapper that could be injected.

I see your point that a default could be instantiated and made overridable for testing only. The reason I added three new autoconfigured beans to `GcpContextAutoConfiguration` is for consistency, since these wrappers behave similarly to `GcpProjectIdProvider`. But if they are not useful for other spring-cloud-gcp projects, I can remove them in favor of overridable defaults.

Reply:
You can use `@SpyBean` or even `@MockBean` for the `ComputeEngineAudienceProvider` and `AppEngineAudienceProvider`: https://docs.spring.io/spring-boot/docs/2.1.0.RELEASE/reference/htmlsingle/#boot-features-testing-spring-boot-applications-mocking-beans.


Otherwise I guess you can just override those beans in the test `@Configuration` with `@Primary`.

Reply:
The behavior change would need to happen after the specific AudienceProvider bean is created, but before AudienceValidator is initialized because `SpyPostProcessor` overrides `postProcessAfterInitialization`, so I think it would run too late in the lifecycle.
Both overriding in @Configuration and using @MockBean would change behavior of autoconfiguration under test.

Do you mind if I leave the current implementation?

Reply:
OK. If there is no other choice...